src/modules/transaction/core/use-cases/to-review/cancel-transaction.use-case.ts

```
@@ -1,16 +1,36 @@
 import { EntityId } from 'src/core/domain/entity/entity-id';
+import { IUseCase } from 'src/core/domain/use-cases/use-case.interface';
 import { TransactionStatus } from 'src/utils/enums';
+import { TransactionOwnershipError } from '../../errors/transaction-ownership-error.error';
 import { ITransactionRepository } from '../../infra/transaction.repository';
 
 export interface CancelTransactionDto {
   transactionId: string;
+  userId: string;
 }
 
-export class CancelTransactionStatusUseCase {
+export class CancelTransactionStatusUseCase implements IUseCase<CancelTransactionDto, boolean> {
   constructor(private readonly transactionRepository: ITransactionRepository) {}
 
+  private isTransactionStatusChangeable(status: TransactionStatus): boolean {
+    return [
+      TransactionStatus.COMPLETED,
+      TransactionStatus.FAILED,
+      TransactionStatus.CANCELLED,
+      TransactionStatus.EXPIRED,
+      TransactionStatus.IN_PROGRESS,
+    ].includes(status);
+  }
+
   public async execute(input: CancelTransactionDto) {
     const transaction = await this.transactionRepository.getById(EntityId.fromString(input.transactionId));
+
+    // Check if transaction belongs to currently logged in user and throw error if it doesn't.
+    if (transaction.userId?.inner !== EntityId.fromString(input.userId).inner) throw new TransactionOwnershipError();
+
+    // Only change status if status is not COMPLETED/FAILED/CANCELLED/EXPIRED
+    if (this.isTransactionStatusChangeable(transaction.status)) return false;
```

You should change this here as you are overcomplicating the logic. If a transaction can be canceled only when the status is `TransactionStatus.PENDING` then simply compare `transaction.status === TransactionStatus.PENDING`, also instead of simply returning false you should return an error as it is more meaningful than simply returning false which does not tell anything meaningful to the user about the result of the operation.